### PR TITLE
Error component

### PR DESCRIPTION
**Pull request by thomasdigby**
#### What have you done
My my, this is fairly nasty, but it _works_ for now. We're now catching errored responses from `AsyncProps` and `react-router` and rendering an `Error` component.

#### Why have you done it
We needed the ability to handle error states

Don't merge this in, this PR is for comments.

### CODE REVIEW

File: src/server.js [hunk 1]
```diff
@@ -132,61 +128,47 @@ export default class Tapestry {
...
+                assets: this.env === 'production' ? this.assets : null
+              })
+            ).code(404)
+          }
 
           // 500 if error from Router
-          if (error)
-            return reply(error.message).code(500)
+          if (err) {
+            error(err)
```
Comment (mathewhawley) [inline src/server.js]: Will the `process.exit(0)` in these error calls kill the server? Will the return statement run?

**Comment by mathewhawley** [comment]
Cool – I don't fully understand Hapi or `async-props`, so a little hard to comment, the only one I would make is a general one – the loaders have a lot of duplication – they maaaaay be a candidate for a HOC? Don't know of the top of my head how that would happen without fiddling around with it

File: src/server.js (hunk 1 above)
Comment (thomasdigby) [inline src/server.js]: That's true, it will kill the server.

**Comment by thomasdigby** [comment]
Yep definitely, unfortunately, I don't think we're able to the way we're using AsyncProps. This probably needs a big 'ol rewrite when we rethink the data loading.

File: src/server.js (hunk 1 above)
Comment (mathewhawley) [inline src/server.js]: I guess it works if it's rendering error components on the client

File: src/render.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    data = {
+      markup: renderStaticOptimized(() =>
+        renderToString(
+          <Error />
+        )
+      ),
+      head: Helmet.rewind(),
+      assets
+    }
+  } else {
```
Comment (thomasdigby) [inline src/render.js]: Ah beautiful, I shall.

**Comment by thomasdigby** [comment]
Yep can do, getting it working is the first part, making it _not terrible_ can come later?

**Pull request merged**